ql/src/test/queries/clientpositive/udf_date_format.q

```diff
@@ -78,3 +78,16 @@ select date_format("2015-04-08 10:30:45","yyyy-MM-dd HH:mm:ss.SSS z");
 --julian date
 set hive.local.time.zone=UTC;
 select date_format("1001-01-05","dd---MM--yyyy");
+
+--dates prior to 1900
+set hive.local.time.zone=Asia/Bangkok;
+select date_format('1400-01-14 01:01:10.123', 'yyyy-MM-dd HH:mm:ss.SSS z');
+select date_format('1800-01-14 01:01:10.123', 'yyyy-MM-dd HH:mm:ss.SSS z');
+
+set hive.local.time.zone=Europe/Berlin;
+select date_format('1400-01-14 01:01:10.123', 'yyyy-MM-dd HH:mm:ss.SSS z');
+select date_format('1800-01-14 01:01:10.123', 'yyyy-MM-dd HH:mm:ss.SSS z');
+
+set hive.local.time.zone=Africa/Johannesburg;
+select date_format('1400-01-14 01:01:10.123', 'yyyy-MM-dd HH:mm:ss.SSS z');
```

Comment:
Do we have tests where the input to `date_format` is a timestamp and not a string literal. In that case what should be the result of `date_format`? 

Should it change when the local timezone changes or it should always be the same? I am asking cause if I remember well the data type that is used when we store timestamps in tables is `TIMESTAMP WITHOUT TIMEZONE` so in principle equivalent to `LocalDateTime`. From a Java perspective (using `DateTimeFormatter`) requesting formatting with timezone from a timestamp without timezone would be an error.

Reply:
We have tests where we use cast function to get Timestamp as input data type.

Reply:
+1 for @zabetak comment. For that we need to add the timeZone  to localDateTime before calling the formatter.

Reply:
@guptanikhil007 
I think, @zabetak point is what will be the interpretation of timezone for a timestamp value stored in a table. Will it be treated as UTC or local timezone? If local timezone, then will the output changes when we change the local timezone config?
From the code, it seems `yes`. But, I guess, the behavior is same with `SimpleDateFormat `as well. Pls confirm.

Reply:
All the timestamps are considered to be in local time zone in both Hive 1.2 and Hive master (patched).
Both Hive 1.2 and 3.1 have similar results for the date_format function.
E.g.: 
Hive 1.2.1 (with timezone set to Asia/Bangkok)
```
0: jdbc:hive2://zk0-nikhil.ae4yqb3genuuvaozdf> select * from test_tbl_orc;
+------------------------------+--------------------------+--+
| test_tbl_orc.clusterversion  |    test_tbl_orc.col2     |
+------------------------------+--------------------------+--+
| Hive 4.0                     | 1800-01-14 01:01:10.123  |
+------------------------------+--------------------------+--+
1 row selected (0.372 seconds)
0: jdbc:hive2://zk0-nikhil.ae4yqb3genuuvaozdf> desc test_tbl_orc;
+-----------------+------------+----------+--+
|    col_name     | data_type  | comment  |
+-----------------+------------+----------+--+
| clusterversion  | string     |          |
| col2            | timestamp  |          |
+-----------------+------------+----------+--+
2 rows selected (0.65 seconds)
0: jdbc:hive2://zk0-nikhil.ae4yqb3genuuvaozdf> select date_format(col2, "yyyy-MM-dd HH:mm:ss.SSS z") from test_tbl_orc;
+------------------------------+--+
|             _c0              |
+------------------------------+--+
| 1800-01-14 01:01:10.123 ICT  |
+------------------------------+--+
```
Hive master (using MiniHS2 and same orc file from Hive 1.2)
```
0: jdbc:hive2://localhost:10000/> select * from test_tbl_orc;
+------------------------------+--------------------------+
| test_tbl_orc.clusterversion  |    test_tbl_orc.col2     |
+------------------------------+--------------------------+
| Hive 4.0                     | 1800-01-14 01:01:10.123  |
+------------------------------+--------------------------+
1 row selected (0.23 seconds)
0: jdbc:hive2://localhost:10000/> desc test_tbl_orc;
+-----------------+------------+----------+
|    col_name     | data_type  | comment  |
+-----------------+------------+----------+
| clusterversion  | string     |          |
| col2            | timestamp  |          |
+-----------------+------------+----------+
2 rows selected (0.24 seconds)
0: jdbc:hive2://localhost:10000/> set hive.local.time.zone=Asia/Bangkok;
No rows affected (0.102 seconds)
0: jdbc:hive2://localhost:10000/> select date_format(col2, "yyyy-MM-dd HH:mm:ss.SSS z") from test_tbl_orc;
+------------------------------+
|             _c0              |
+------------------------------+
| 1800-01-14 01:01:10.123 ICT  |
+------------------------------+
1 row selected (0.261 seconds)
0: jdbc:hive2://localhost:10000/> set hive.local.time.zone=LOCAL;
No rows affected (0.036 seconds)
0: jdbc:hive2://localhost:10000/> select date_format(col2, "yyyy-MM-dd HH:mm:ss.SSS z") from test_tbl_orc;
+------------------------------+
|             _c0              |
+------------------------------+
| 1800-01-14 01:01:10.123 PST  |
+------------------------------+
1 row selected (0.492 seconds)
```

Reply:
Thanks for testing this @guptanikhil007 .